In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.DataFrame({
    'salary':np.random.randint(30, 90, 20),
    'height':np.random.randint(120, 190, 20),
    'weight':np.random.randint(40, 90, 20),
    'age': np.random.randint(20, 35, 20),
    'country':np.random.choice(['india', 'usa', 'aus', 'china'], 20),
    'gender': np.random.choice(['male', 'female'], 20)
}, index=[chr(i) for i in range(65, 65+20)])
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,female
B,37,123,89,32,china,female
C,84,153,52,31,usa,male
D,82,182,67,30,china,female
E,49,154,61,31,aus,female
F,62,177,84,34,india,male
G,43,183,79,28,china,female
H,50,189,75,24,india,male
I,32,142,78,31,usa,male
J,71,121,84,32,aus,male


## Apply

In [3]:
df['height']

A    126
B    123
C    153
D    182
E    154
F    177
G    183
H    189
I    142
J    121
K    129
L    177
M    168
N    164
O    145
P    145
Q    143
R    124
S    148
T    145
Name: height, dtype: int32

In [4]:
df['height'].apply(lambda x: f"{x} cm")


A    126 cm
B    123 cm
C    153 cm
D    182 cm
E    154 cm
F    177 cm
G    183 cm
H    189 cm
I    142 cm
J    121 cm
K    129 cm
L    177 cm
M    168 cm
N    164 cm
O    145 cm
P    145 cm
Q    143 cm
R    124 cm
S    148 cm
T    145 cm
Name: height, dtype: object

In [5]:
df['height'] = df['height'].apply(lambda x: f"{x} cm")

In [6]:
df

,salary,height,weight,age,country,gender
A,68,126 cm,50,30,china,female
B,37,123 cm,89,32,china,female
C,84,153 cm,52,31,usa,male
D,82,182 cm,67,30,china,female
E,49,154 cm,61,31,aus,female
F,62,177 cm,84,34,india,male
G,43,183 cm,79,28,china,female
H,50,189 cm,75,24,india,male
I,32,142 cm,78,31,usa,male
J,71,121 cm,84,32,aus,male


In [7]:
for i in df['height']:
    print(f"{i} cm")

126 cm cm
123 cm cm
153 cm cm
182 cm cm
154 cm cm
177 cm cm
183 cm cm
189 cm cm
142 cm cm
121 cm cm
129 cm cm
177 cm cm
168 cm cm
164 cm cm
145 cm cm
145 cm cm
143 cm cm
124 cm cm
148 cm cm
145 cm cm


In [8]:
df['height']

A    126 cm
B    123 cm
C    153 cm
D    182 cm
E    154 cm
F    177 cm
G    183 cm
H    189 cm
I    142 cm
J    121 cm
K    129 cm
L    177 cm
M    168 cm
N    164 cm
O    145 cm
P    145 cm
Q    143 cm
R    124 cm
S    148 cm
T    145 cm
Name: height, dtype: object

In [9]:
df

,salary,height,weight,age,country,gender
A,68,126 cm,50,30,china,female
B,37,123 cm,89,32,china,female
C,84,153 cm,52,31,usa,male
D,82,182 cm,67,30,china,female
E,49,154 cm,61,31,aus,female
F,62,177 cm,84,34,india,male
G,43,183 cm,79,28,china,female
H,50,189 cm,75,24,india,male
I,32,142 cm,78,31,usa,male
J,71,121 cm,84,32,aus,male


In [10]:
df['height'] = df['height'].apply(lambda x: int(x.split()[0]))

In [11]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,female
B,37,123,89,32,china,female
C,84,153,52,31,usa,male
D,82,182,67,30,china,female
E,49,154,61,31,aus,female
F,62,177,84,34,india,male
G,43,183,79,28,china,female
H,50,189,75,24,india,male
I,32,142,78,31,usa,male
J,71,121,84,32,aus,male


In [12]:
df['gender'].apply(lambda x: x.upper())

A    FEMALE
B    FEMALE
C      MALE
D    FEMALE
E    FEMALE
F      MALE
G    FEMALE
H      MALE
I      MALE
J      MALE
K      MALE
L      MALE
M      MALE
N    FEMALE
O    FEMALE
P    FEMALE
Q      MALE
R      MALE
S      MALE
T      MALE
Name: gender, dtype: object

In [13]:
def func(x):
    return x.upper()

In [14]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,female
B,37,123,89,32,china,female
C,84,153,52,31,usa,male
D,82,182,67,30,china,female
E,49,154,61,31,aus,female
F,62,177,84,34,india,male
G,43,183,79,28,china,female
H,50,189,75,24,india,male
I,32,142,78,31,usa,male
J,71,121,84,32,aus,male


In [15]:
df['gender'] = df['gender'].map({'male':'M', 'female':'F'})

In [16]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,F
B,37,123,89,32,china,F
C,84,153,52,31,usa,M
D,82,182,67,30,china,F
E,49,154,61,31,aus,F
F,62,177,84,34,india,M
G,43,183,79,28,china,F
H,50,189,75,24,india,M
I,32,142,78,31,usa,M
J,71,121,84,32,aus,M


In [19]:
df[df['country'] == 'india']

,salary,height,weight,age,country,gender
F,62,177,84,34,india,M
H,50,189,75,24,india,M
K,61,129,81,22,india,M
L,66,177,46,33,india,M
M,85,168,61,20,india,M
N,87,164,85,31,india,F
Q,79,143,77,26,india,M
S,73,148,56,26,india,M


In [22]:
df[(df['country'] == 'india') & (df['gender'] == 'M')]


,salary,height,weight,age,country,gender
F,62,177,84,34,india,M
H,50,189,75,24,india,M
K,61,129,81,22,india,M
L,66,177,46,33,india,M
M,85,168,61,20,india,M
Q,79,143,77,26,india,M
S,73,148,56,26,india,M


In [21]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,F
B,37,123,89,32,china,F
C,84,153,52,31,usa,M
D,82,182,67,30,china,F
E,49,154,61,31,aus,F
F,62,177,84,34,india,M
G,43,183,79,28,china,F
H,50,189,75,24,india,M
I,32,142,78,31,usa,M
J,71,121,84,32,aus,M


In [23]:
df[df['country'] == 'india']['salary'].mean()

np.float64(70.375)

In [24]:
df[(df['country'] == 'usa') & (df['gender'] == 'M')]['salary'].mean()

np.float64(63.75)

In [25]:
df[(df['country'] == 'usa') & (df['gender'] == 'F')]['salary'].mean()


nan

In [26]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,F
B,37,123,89,32,china,F
C,84,153,52,31,usa,M
D,82,182,67,30,china,F
E,49,154,61,31,aus,F
F,62,177,84,34,india,M
G,43,183,79,28,china,F
H,50,189,75,24,india,M
I,32,142,78,31,usa,M
J,71,121,84,32,aus,M


In [27]:
df.groupby('country')

In [29]:
df.groupby('country').groups


{'aus': ['E', 'J', 'P'], 'china': ['A', 'B', 'D', 'G', 'O'], 'india': ['F', 'H', 'K', 'L', 'M', 'N', 'Q', 'S'], 'usa': ['C', 'I', 'R', 'T']}

In [30]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,F
B,37,123,89,32,china,F
C,84,153,52,31,usa,M
D,82,182,67,30,china,F
E,49,154,61,31,aus,F
F,62,177,84,34,india,M
G,43,183,79,28,china,F
H,50,189,75,24,india,M
I,32,142,78,31,usa,M
J,71,121,84,32,aus,M


In [31]:
df.groupby('country')['salary'].mean()


country
aus      69.666667
china    60.800000
india    70.375000
usa      63.750000
Name: salary, dtype: float64

In [32]:
df.groupby('country')[['salary', 'weight', 'height']].mean()


,salary,weight,height
country,,,
aus,69.666667,75.666667,140.000
china,60.800000,71.000000,151.800
india,70.375000,70.625000,161.875
usa,63.750000,61.750000,141.000


In [33]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,F
B,37,123,89,32,china,F
C,84,153,52,31,usa,M
D,82,182,67,30,china,F
E,49,154,61,31,aus,F
F,62,177,84,34,india,M
G,43,183,79,28,china,F
H,50,189,75,24,india,M
I,32,142,78,31,usa,M
J,71,121,84,32,aus,M


In [ ]:
df.groupby('country').agg({ # This code groups the DataFrame df by the 'country' column and then applies aggregation functions to the 'salary' and 'weight' columns. Specifically, it calculates the mean of the 'salary' column and both the mean and median of the 'weight' column for each country group. The result is a new DataFrame that summarizes these statistics for each country.
    'salary':'mean',
    'weight':['mean', 'median']
})


salary     weight       
              mean       mean median
country                             
aus      69.666667  75.666667   82.0
china    60.800000  71.000000   70.0
india    70.375000  70.625000   76.0
usa      63.750000  61.750000   60.0

In [35]:
df.groupby(['country', 'gender'])['salary'].mean()

country  gender
aus      F         69.00
         M         71.00
china    F         60.80
india    F         87.00
         M         68.00
usa      M         63.75
Name: salary, dtype: float64

In [36]:
df.groupby(['country', 'gender']).agg({
    'salary':'mean',
    'height':'median',
    'age':['mean', 'median'],
})


salary height        age       
                 mean median       mean median
country gender                                
aus     F       69.00  149.5  26.000000   26.0
        M       71.00  121.0  32.000000   32.0
china   F       60.80  145.0  28.200000   30.0
india   F       87.00  164.0  31.000000   31.0
        M       68.00  168.0  26.428571   26.0
usa     M       63.75  143.5  32.000000   32.0

In [37]:
df.groupby(['country', 'gender']).agg({
    'salary':'mean',
    'height':'median',
    'age':['mean', 'median'],
    'gender':'count'
})


salary height        age        gender
                 mean median       mean median  count
country gender                                       
aus     F       69.00  149.5  26.000000   26.0      2
        M       71.00  121.0  32.000000   32.0      1
china   F       60.80  145.0  28.200000   30.0      5
india   F       87.00  164.0  31.000000   31.0      1
        M       68.00  168.0  26.428571   26.0      7
usa     M       63.75  143.5  32.000000   32.0      4

In [38]:
df.groupby(['country'])

In [39]:
idx = df.groupby(['country']).groups
idx

{'aus': ['E', 'J', 'P'], 'china': ['A', 'B', 'D', 'G', 'O'], 'india': ['F', 'H', 'K', 'L', 'M', 'N', 'Q', 'S'], 'usa': ['C', 'I', 'R', 'T']}

In [40]:
idx['aus']

Index(['E', 'J', 'P'], dtype='object')

In [41]:
df.loc[idx['aus']]

,salary,height,weight,age,country,gender
E,49,154,61,31,aus,F
J,71,121,84,32,aus,M
P,89,145,82,21,aus,F


In [42]:
df.loc[idx['india']]


,salary,height,weight,age,country,gender
F,62,177,84,34,india,M
H,50,189,75,24,india,M
K,61,129,81,22,india,M
L,66,177,46,33,india,M
M,85,168,61,20,india,M
N,87,164,85,31,india,F
Q,79,143,77,26,india,M
S,73,148,56,26,india,M


In [43]:
df1 = df.groupby(['country', 'gender']).agg({
    'salary':'mean',
    'height':'median',
    'age':['mean', 'median'],
    'gender':'count'
})
df1

salary height        age        gender
                 mean median       mean median  count
country gender                                       
aus     F       69.00  149.5  26.000000   26.0      2
        M       71.00  121.0  32.000000   32.0      1
china   F       60.80  145.0  28.200000   30.0      5
india   F       87.00  164.0  31.000000   31.0      1
        M       68.00  168.0  26.428571   26.0      7
usa     M       63.75  143.5  32.000000   32.0      4

In [44]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,F
B,37,123,89,32,china,F
C,84,153,52,31,usa,M
D,82,182,67,30,china,F
E,49,154,61,31,aus,F
F,62,177,84,34,india,M
G,43,183,79,28,china,F
H,50,189,75,24,india,M
I,32,142,78,31,usa,M
J,71,121,84,32,aus,M


In [45]:
df.index

Index(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N',
       'O', 'P', 'Q', 'R', 'S', 'T'],
      dtype='object')

In [46]:
df.columns

Index(['salary', 'height', 'weight', 'age', 'country', 'gender'], dtype='object')

In [47]:
df1

salary height        age        gender
                 mean median       mean median  count
country gender                                       
aus     F       69.00  149.5  26.000000   26.0      2
        M       71.00  121.0  32.000000   32.0      1
china   F       60.80  145.0  28.200000   30.0      5
india   F       87.00  164.0  31.000000   31.0      1
        M       68.00  168.0  26.428571   26.0      7
usa     M       63.75  143.5  32.000000   32.0      4

In [48]:
df1.index

MultiIndex([(  'aus', 'F'),
            (  'aus', 'M'),
            ('china', 'F'),
            ('india', 'F'),
            ('india', 'M'),
            (  'usa', 'M')],
           names=['country', 'gender'])

In [49]:
print(dir(pd))

['ArrowDtype', 'BooleanDtype', 'Categorical', 'CategoricalDtype', 'CategoricalIndex', 'DataFrame', 'DateOffset', 'DatetimeIndex', 'DatetimeTZDtype', 'ExcelFile', 'ExcelWriter', 'Flags', 'Float32Dtype', 'Float64Dtype', 'Grouper', 'HDFStore', 'Index', 'IndexSlice', 'Int16Dtype', 'Int32Dtype', 'Int64Dtype', 'Int8Dtype', 'Interval', 'IntervalDtype', 'IntervalIndex', 'MultiIndex', 'NA', 'NaT', 'NamedAgg', 'Period', 'PeriodDtype', 'PeriodIndex', 'RangeIndex', 'Series', 'SparseDtype', 'StringDtype', 'Timedelta', 'TimedeltaIndex', 'Timestamp', 'UInt16Dtype', 'UInt32Dtype', 'UInt64Dtype', 'UInt8Dtype', '__all__', '__builtins__', '__cached__', '__doc__', '__docformat__', '__file__', '__git_version__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '__version__', '_built_with_meson', '_config', '_is_numpy_dev', '_libs', '_pandas_datetime_CAPI', '_pandas_parser_CAPI', '_testing', '_typing', '_version_meson', 'annotations', 'api', 'array', 'arrays', 'bdate_range', 'compat', 'conca

In [50]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,F
B,37,123,89,32,china,F
C,84,153,52,31,usa,M
D,82,182,67,30,china,F
E,49,154,61,31,aus,F
F,62,177,84,34,india,M
G,43,183,79,28,china,F
H,50,189,75,24,india,M
I,32,142,78,31,usa,M
J,71,121,84,32,aus,M


In [60]:
print(dir(df))

['T', '_AXIS_LEN', '_AXIS_ORDERS', '_AXIS_TO_AXIS_NUMBER', '_HANDLED_TYPES', '__abs__', '__add__', '__and__', '__annotations__', '__array__', '__array_priority__', '__array_ufunc__', '__bool__', '__class__', '__contains__', '__copy__', '__dataframe__', '__dataframe_consortium_standard__', '__deepcopy__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__divmod__', '__doc__', '__eq__', '__finalize__', '__floordiv__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__iadd__', '__iand__', '__ifloordiv__', '__imod__', '__imul__', '__init__', '__init_subclass__', '__invert__', '__ior__', '__ipow__', '__isub__', '__iter__', '__itruediv__', '__ixor__', '__le__', '__len__', '__lt__', '__matmul__', '__mod__', '__module__', '__mul__', '__ne__', '__neg__', '__new__', '__nonzero__', '__or__', '__pandas_priority__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloord

In [51]:
len(dir(df))


448

In [52]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,F
B,37,123,89,32,china,F
C,84,153,52,31,usa,M
D,82,182,67,30,china,F
E,49,154,61,31,aus,F
F,62,177,84,34,india,M
G,43,183,79,28,china,F
H,50,189,75,24,india,M
I,32,142,78,31,usa,M
J,71,121,84,32,aus,M


In [53]:
print(df.to_dict())

{'salary': {'A': 68, 'B': 37, 'C': 84, 'D': 82, 'E': 49, 'F': 62, 'G': 43, 'H': 50, 'I': 32, 'J': 71, 'K': 61, 'L': 66, 'M': 85, 'N': 87, 'O': 74, 'P': 89, 'Q': 79, 'R': 50, 'S': 73, 'T': 89}, 'height': {'A': 126, 'B': 123, 'C': 153, 'D': 182, 'E': 154, 'F': 177, 'G': 183, 'H': 189, 'I': 142, 'J': 121, 'K': 129, 'L': 177, 'M': 168, 'N': 164, 'O': 145, 'P': 145, 'Q': 143, 'R': 124, 'S': 148, 'T': 145}, 'weight': {'A': 50, 'B': 89, 'C': 52, 'D': 67, 'E': 61, 'F': 84, 'G': 79, 'H': 75, 'I': 78, 'J': 84, 'K': 81, 'L': 46, 'M': 61, 'N': 85, 'O': 70, 'P': 82, 'Q': 77, 'R': 68, 'S': 56, 'T': 49}, 'age': {'A': 30, 'B': 32, 'C': 31, 'D': 30, 'E': 31, 'F': 34, 'G': 28, 'H': 24, 'I': 31, 'J': 32, 'K': 22, 'L': 33, 'M': 20, 'N': 31, 'O': 21, 'P': 21, 'Q': 26, 'R': 33, 'S': 26, 'T': 33}, 'country': {'A': 'china', 'B': 'china', 'C': 'usa', 'D': 'china', 'E': 'aus', 'F': 'india', 'G': 'china', 'H': 'india', 'I': 'usa', 'J': 'aus', 'K': 'india', 'L': 'india', 'M': 'india', 'N': 'india', 'O': 'china', 

In [54]:
df.to_html()

'<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>salary</th>\n      <th>height</th>\n      <th>weight</th>\n      <th>age</th>\n      <th>country</th>\n      <th>gender</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>A</th>\n      <td>68</td>\n      <td>126</td>\n      <td>50</td>\n      <td>30</td>\n      <td>china</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>B</th>\n      <td>37</td>\n      <td>123</td>\n      <td>89</td>\n      <td>32</td>\n      <td>china</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>C</th>\n      <td>84</td>\n      <td>153</td>\n      <td>52</td>\n      <td>31</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>D</th>\n      <td>82</td>\n      <td>182</td>\n      <td>67</td>\n      <td>30</td>\n      <td>china</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>E</th>\n      <td>49</td>\n      <td>154</td>\n      <td>61</td>\n      <td>31</td>\

'<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>salary</th>\n      <th>height</th>\n      <th>weight</th>\n      <th>age</th>\n      <th>country</th>\n      <th>gender</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>A</th>\n      <td>68</td>\n      <td>126</td>\n      <td>50</td>\n      <td>30</td>\n      <td>china</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>B</th>\n      <td>37</td>\n      <td>123</td>\n      <td>89</td>\n      <td>32</td>\n      <td>china</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>C</th>\n      <td>84</td>\n      <td>153</td>\n      <td>52</td>\n      <td>31</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>D</th>\n      <td>82</td>\n      <td>182</td>\n      <td>67</td>\n      <td>30</td>\n      <td>china</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>E</th>\n      <td>49</td>\n      <td>154</td>\n      <td>61</td>\n      <td>31</td>\n      <td>aus</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>F</th>\n      <td>62</td>\n      <td>177</td>\n      <td>84</td>\n      <td>34</td>\n      <td>india</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>G</th>\n      <td>43</td>\n      <td>183</td>\n      <td>79</td>\n      <td>28</td>\n      <td>china</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>H</th>\n      <td>50</td>\n      <td>189</td>\n      <td>75</td>\n      <td>24</td>\n      <td>india</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>I</th>\n      <td>32</td>\n      <td>142</td>\n      <td>78</td>\n      <td>31</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>J</th>\n      <td>71</td>\n      <td>121</td>\n      <td>84</td>\n      <td>32</td>\n      <td>aus</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>K</th>\n      <td>61</td>\n      <td>129</td>\n      <td>81</td>\n      <td>22</td>\n      <td>india</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>L</th>\n      <td>66</td>\n      <td>177</td>\n      <td>46</td>\n      <td>33</td>\n      <td>india</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>M</th>\n      <td>85</td>\n      <td>168</td>\n      <td>61</td>\n      <td>20</td>\n      <td>india</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>N</th>\n      <td>87</td>\n      <td>164</td>\n      <td>85</td>\n      <td>31</td>\n      <td>india</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>O</th>\n      <td>74</td>\n      <td>145</td>\n      <td>70</td>\n      <td>21</td>\n      <td>china</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>P</th>\n      <td>89</td>\n      <td>145</td>\n      <td>82</td>\n      <td>21</td>\n      <td>aus</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>Q</th>\n      <td>79</td>\n      <td>143</td>\n      <td>77</td>\n      <td>26</td>\n      <td>india</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>R</th>\n      <td>50</td>\n      <td>124</td>\n      <td>68</td>\n      <td>33</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>S</th>\n      <td>73</td>\n      <td>148</td>\n      <td>56</td>\n      <td>26</td>\n      <td>india</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>T</th>\n      <td>89</td>\n      <td>145</td>\n      <td>49</td>\n      <td>33</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n  </tbody>\n</table>'


In [55]:
df.to_numpy()

array([[68, 126, 50, 30, 'china', 'F'],
       [37, 123, 89, 32, 'china', 'F'],
       [84, 153, 52, 31, 'usa', 'M'],
       [82, 182, 67, 30, 'china', 'F'],
       [49, 154, 61, 31, 'aus', 'F'],
       [62, 177, 84, 34, 'india', 'M'],
       [43, 183, 79, 28, 'china', 'F'],
       [50, 189, 75, 24, 'india', 'M'],
       [32, 142, 78, 31, 'usa', 'M'],
       [71, 121, 84, 32, 'aus', 'M'],
       [61, 129, 81, 22, 'india', 'M'],
       [66, 177, 46, 33, 'india', 'M'],
       [85, 168, 61, 20, 'india', 'M'],
       [87, 164, 85, 31, 'india', 'F'],
       [74, 145, 70, 21, 'china', 'F'],
       [89, 145, 82, 21, 'aus', 'F'],
       [79, 143, 77, 26, 'india', 'M'],
       [50, 124, 68, 33, 'usa', 'M'],
       [73, 148, 56, 26, 'india', 'M'],
       [89, 145, 49, 33, 'usa', 'M']], dtype=object)

In [56]:
df1

salary height        age        gender
                 mean median       mean median  count
country gender                                       
aus     F       69.00  149.5  26.000000   26.0      2
        M       71.00  121.0  32.000000   32.0      1
china   F       60.80  145.0  28.200000   30.0      5
india   F       87.00  164.0  31.000000   31.0      1
        M       68.00  168.0  26.428571   26.0      7
usa     M       63.75  143.5  32.000000   32.0      4

In [57]:
df

,salary,height,weight,age,country,gender
A,68,126,50,30,china,F
B,37,123,89,32,china,F
C,84,153,52,31,usa,M
D,82,182,67,30,china,F
E,49,154,61,31,aus,F
F,62,177,84,34,india,M
G,43,183,79,28,china,F
H,50,189,75,24,india,M
I,32,142,78,31,usa,M
J,71,121,84,32,aus,M


In [58]:
df1

salary height        age        gender
                 mean median       mean median  count
country gender                                       
aus     F       69.00  149.5  26.000000   26.0      2
        M       71.00  121.0  32.000000   32.0      1
china   F       60.80  145.0  28.200000   30.0      5
india   F       87.00  164.0  31.000000   31.0      1
        M       68.00  168.0  26.428571   26.0      7
usa     M       63.75  143.5  32.000000   32.0      4

In [63]:
df1.stack() # This code converts the DataFrame df1 from wide format to long format by stacking the columns into rows.

C:\Users\Mayan\AppData\Local\Temp\ipykernel_3552\4255760901.py:1: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.
  df1.stack() # This code converts the DataFrame df1 from wide format to long format by stacking the columns into rows.


salary  height        age  gender
country gender                                          
aus     F      count      NaN     NaN        NaN     2.0
               mean     69.00     NaN  26.000000     NaN
               median     NaN   149.5  26.000000     NaN
        M      count      NaN     NaN        NaN     1.0
               mean     71.00     NaN  32.000000     NaN
               median     NaN   121.0  32.000000     NaN
china   F      count      NaN     NaN        NaN     5.0
               mean     60.80     NaN  28.200000     NaN
               median     NaN   145.0  30.000000     NaN
india   F      count      NaN     NaN        NaN     1.0
               mean     87.00     NaN  31.000000     NaN
               median     NaN   164.0  31.000000     NaN
        M      count      NaN     NaN        NaN     7.0
               mean     68.00     NaN  26.428571     NaN
               median     NaN   168.0  26.000000     NaN
usa     M      count      NaN     NaN        NaN     4.0
               mean     63.75     NaN  32.000000     NaN
               median     NaN   143.5  32.000000     NaN

In [64]:
df1

salary height        age        gender
                 mean median       mean median  count
country gender                                       
aus     F       69.00  149.5  26.000000   26.0      2
        M       71.00  121.0  32.000000   32.0      1
china   F       60.80  145.0  28.200000   30.0      5
india   F       87.00  164.0  31.000000   31.0      1
        M       68.00  168.0  26.428571   26.0      7
usa     M       63.75  143.5  32.000000   32.0      4

In [65]:
df1.stack(level=0) # This code stacks the DataFrame df1 along the first level of its column index, effectively converting the DataFrame from wide format to long format by moving the first level of column labels into the row index.

C:\Users\Mayan\AppData\Local\Temp\ipykernel_3552\4242375300.py:1: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.
  df1.stack(level=0) # This code stacks the DataFrame df1 along the first level of its column index, effectively converting the DataFrame from wide format to long format by moving the first level of column labels into the row index.


mean  median  count
country gender                                 
aus     F      age     26.000000    26.0    NaN
               gender        NaN     NaN    2.0
               height        NaN   149.5    NaN
               salary  69.000000     NaN    NaN
        M      age     32.000000    32.0    NaN
               gender        NaN     NaN    1.0
               height        NaN   121.0    NaN
               salary  71.000000     NaN    NaN
china   F      age     28.200000    30.0    NaN
               gender        NaN     NaN    5.0
               height        NaN   145.0    NaN
               salary  60.800000     NaN    NaN
india   F      age     31.000000    31.0    NaN
               gender        NaN     NaN    1.0
               height        NaN   164.0    NaN
               salary  87.000000     NaN    NaN
        M      age     26.428571    26.0    NaN
               gender        NaN     NaN    7.0
               height        NaN   168.0    NaN
               salary  68.000000     NaN    NaN
usa     M      age     32.000000    32.0    NaN
               gender        NaN     NaN    4.0
               height        NaN   143.5    NaN
               salary  63.750000     NaN    NaN

In [66]:
df1

salary height        age        gender
                 mean median       mean median  count
country gender                                       
aus     F       69.00  149.5  26.000000   26.0      2
        M       71.00  121.0  32.000000   32.0      1
china   F       60.80  145.0  28.200000   30.0      5
india   F       87.00  164.0  31.000000   31.0      1
        M       68.00  168.0  26.428571   26.0      7
usa     M       63.75  143.5  32.000000   32.0      4

In [67]:
df1.unstack()

salary        height          age                         gender     
          mean        median         mean            median        count     
gender       F      M      F      M     F          M      F     M      F    M
country                                                                      
aus       69.0  71.00  149.5  121.0  26.0  32.000000   26.0  32.0    2.0  1.0
china     60.8    NaN  145.0    NaN  28.2        NaN   30.0   NaN    5.0  NaN
india     87.0  68.00  164.0  168.0  31.0  26.428571   31.0  26.0    1.0  7.0
usa        NaN  63.75    NaN  143.5   NaN  32.000000    NaN  32.0    NaN  4.0

In [68]:
df1.unstack().unstack()


                gender  country
salary  mean    F       aus         69.000000
                        china       60.800000
                        india       87.000000
                        usa               NaN
                M       aus         71.000000
                        china             NaN
                        india       68.000000
                        usa         63.750000
height  median  F       aus        149.500000
                        china      145.000000
                        india      164.000000
                        usa               NaN
                M       aus        121.000000
                        china             NaN
                        india      168.000000
                        usa        143.500000
age     mean    F       aus         26.000000
                        china       28.200000
                        india       31.000000
                        usa               NaN
                M       aus         32.000000
  

In [39]:
data = pd.read_csv('D:/Program/Data Science/DataCollection/MobileData.csv') # This code reads a CSV file located at the specified path and loads its contents into a pandas DataFrame named
data

,Unnamed: 0,Mobile,Ratings,Prices
0,0,"MOTOROLA g37 (PANTONE Nautical Blue, 64 GB)",4.3,"₹14,999"
1,1,MOTOROLA g57 power 5G Snapdragon (Pantone Rega...,4.4,"₹20,999"
2,2,"MOTOROLA Edge 60 Pro (Pantone Shadow, 256 GB)",4.3,"₹29,999"
3,3,"MOTOROLA g37 (PANTONE Impenetrable, 64 GB)",4.3,"₹14,999"
4,4,"MOTOROLA Edge 60 Pro (Pantone Shadow, 256 GB)",4.3,"₹33,999"
5,5,MOTOROLA g57 power 5G Snapdragon (Pantone Cors...,4.4,"₹20,999"
6,6,MOTOROLA Edge 60 Fusion 5G (PANTONE Mykonos Bl...,4.4,"₹27,999"
7,7,"MOTOROLA g37 (PANTONE Capri, 64 GB)",4.3,"₹14,999"
8,8,"MOTOROLA Edge 60 Pro (Pantone Sparkling Grape,...",4.3,"₹29,999"
9,9,"MOTOROLA g67 power 5G (Pantone CILANTRO, 256 GB)",4.3,"₹26,999"


In [40]:
data.drop('Unnamed: 0', axis=1, inplace=True)

In [41]:
data

,Mobile,Ratings,Prices
0,"MOTOROLA g37 (PANTONE Nautical Blue, 64 GB)",4.3,"₹14,999"
1,MOTOROLA g57 power 5G Snapdragon (Pantone Rega...,4.4,"₹20,999"
2,"MOTOROLA Edge 60 Pro (Pantone Shadow, 256 GB)",4.3,"₹29,999"
3,"MOTOROLA g37 (PANTONE Impenetrable, 64 GB)",4.3,"₹14,999"
4,"MOTOROLA Edge 60 Pro (Pantone Shadow, 256 GB)",4.3,"₹33,999"
5,MOTOROLA g57 power 5G Snapdragon (Pantone Cors...,4.4,"₹20,999"
6,MOTOROLA Edge 60 Fusion 5G (PANTONE Mykonos Bl...,4.4,"₹27,999"
7,"MOTOROLA g37 (PANTONE Capri, 64 GB)",4.3,"₹14,999"
8,"MOTOROLA Edge 60 Pro (Pantone Sparkling Grape,...",4.3,"₹29,999"
9,"MOTOROLA g67 power 5G (Pantone CILANTRO, 256 GB)",4.3,"₹26,999"


In [42]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Mobile   24 non-null     object 
 1   Ratings  24 non-null     float64
 2   Prices   24 non-null     object 
dtypes: float64(1), object(2)
memory usage: 708.0+ bytes


In [43]:
data

,Mobile,Ratings,Prices
0,"MOTOROLA g37 (PANTONE Nautical Blue, 64 GB)",4.3,"₹14,999"
1,MOTOROLA g57 power 5G Snapdragon (Pantone Rega...,4.4,"₹20,999"
2,"MOTOROLA Edge 60 Pro (Pantone Shadow, 256 GB)",4.3,"₹29,999"
3,"MOTOROLA g37 (PANTONE Impenetrable, 64 GB)",4.3,"₹14,999"
4,"MOTOROLA Edge 60 Pro (Pantone Shadow, 256 GB)",4.3,"₹33,999"
5,MOTOROLA g57 power 5G Snapdragon (Pantone Cors...,4.4,"₹20,999"
6,MOTOROLA Edge 60 Fusion 5G (PANTONE Mykonos Bl...,4.4,"₹27,999"
7,"MOTOROLA g37 (PANTONE Capri, 64 GB)",4.3,"₹14,999"
8,"MOTOROLA Edge 60 Pro (Pantone Sparkling Grape,...",4.3,"₹29,999"
9,"MOTOROLA g67 power 5G (Pantone CILANTRO, 256 GB)",4.3,"₹26,999"
